firedrake/mesh.py

```diff
@@ -2275,14 +2240,26 @@ def __init__(self, coordinates):
             The `CoordinatelessFunction` containing the coordinates.
 
         """
+        import firedrake.functionspaceimpl as functionspaceimpl
+        import firedrake.function as function
+
+        utils._init()
+
+        element = coordinates.ufl_element()
+        uid = utils._new_uid(internal_comm(coordinates.comm, self))
+        super().__init__(element, ufl_id=uid)
+
         topology = coordinates.function_space().mesh()
 
         # this is codegen information so we attach it to the MeshGeometry rather than its cargo
         self.extruded = isinstance(topology, ExtrudedMeshTopology)
         self.variable_layers = self.extruded and topology.variable_layers
 
-        # initialise the mesh cargo
-        self.ufl_cargo().init(coordinates)
+        self.topology = topology
+        self.geometric_shared_data_cache = defaultdict(dict)
+
+        V = functionspaceimpl.WithGeometry.create(coordinates.function_space(), self)
+        self._coordinates = function.Function(V, val=coordinates)
```

Comment:
`self._coordinates` must be a `CoordinatelessFunction`; i.e.,
```suggestion
        self._coordinates = coordinates
```
. `CheckpointFile` is especially strict about the difference between geometry and topology.